// @anon/fern-sdk/src/api/resources/activity/client/Client.ts
/**
 * This file was auto-generated by Fern from our API Definition.
 */

import * as environments from "../../../../environments";
import * as core from "../../../../core";
import * as AnonApi from "../../../index";
import urlJoin from "url-join";
import * as errors from "../../../../errors/index";

export declare namespace Activity {
  interface Options {
    environment?: core.Supplier<environments.AnonApiEnvironment | string>;
    token: core.Supplier<core.BearerToken>;
    /** Override the X-Anon-Sdk-Client-Id header */
    anonSdkClientId: core.Supplier<string>;
  }

  interface RequestOptions {
    /** The maximum time to wait for a response in seconds. */
    timeoutInSeconds?: number;
    /** The number of times to retry the request. Defaults to 2. */
    maxRetries?: number;
    /** A hook to abort the request. */
    abortSignal?: AbortSignal;
    /** Override the X-Anon-Sdk-Client-Id header */
    anonSdkClientId?: string;
    /** Additional headers to include in the request. */
    headers?: Record<string, string>;
  }
}

export class Activity {
  constructor(protected readonly _options: Activity.Options) {}

  /**
   * Retrieve all events for an organization via organizationId. The organizationId is retrieved from the api key via kong. It can also query by both group id and session type.
   *
   * @param {AnonApi.GetActivityEventsRequest} request
   * @param {Activity.RequestOptions} requestOptions - Request-specific configuration.
   *
   * @example
   *     await client.activity.getActivityEvents({
   *         groupId: "string",
   *         activityType: "Link",
   *         app: "string",
   *         startTime: "string",
   *         endTime: "string",
   *         userId: "string",
   *         order: "string"
   *     })
   */
  public async getActivityEvents(
    request: AnonApi.GetActivityEventsRequest = {},
    requestOptions?: Activity.RequestOptions,
  ): Promise<AnonApi.ActivityEventList> {
    const { groupId, activityType, app, startTime, endTime, userId, order } =
      request;
    const _queryParams: Record<string, string | string[] | object | object[]> =
      {};
    if (groupId != null) {
      _queryParams["groupId"] = groupId;
    }

    if (activityType != null) {
      _queryParams["activityType"] = activityType;
    }

    if (app != null) {
      _queryParams["app"] = app;
    }

    if (startTime != null) {
      _queryParams["startTime"] = startTime;
    }

    if (endTime != null) {
      _queryParams["endTime"] = endTime;
    }

    if (userId != null) {
      _queryParams["userId"] = userId;
    }

    if (order != null) {
      _queryParams["order"] = order;
    }

    const _response = await core.fetcher({
      url: urlJoin(
        (await core.Supplier.get(this._options.environment)) ??
          environments.AnonApiEnvironment.Sandbox,
        "/account/api/v1/activity",
      ),
      method: "GET",
      headers: {
        Authorization: await this._getAuthorizationHeader(),
        "X-Anon-Sdk-Client-Id": await core.Supplier.get(
          this._options.anonSdkClientId,
        ),
        "X-Fern-Language": "JavaScript",
        "X-Fern-Runtime": core.RUNTIME.type,
        "X-Fern-Runtime-Version": core.RUNTIME.version,
        ...requestOptions?.headers,
      },
      contentType: "application/json",
      queryParameters: _queryParams,
      requestType: "json",
      timeoutMs:
        requestOptions?.timeoutInSeconds != null
          ? requestOptions.timeoutInSeconds * 1000
          : 60000,
      maxRetries: requestOptions?.maxRetries,
      abortSignal: requestOptions?.abortSignal,
    });
    if (_response.ok) {
      return _response.body as AnonApi.ActivityEventList;
    }

    if (_response.error.reason === "status-code") {
      throw new errors.AnonApiError({
        statusCode: _response.error.statusCode,
        body: _response.error.body,
      });
    }

    switch (_response.error.reason) {
      case "non-json":
        throw new errors.AnonApiError({
          statusCode: _response.error.statusCode,
          body: _response.error.rawBody,
        });
      case "timeout":
        throw new errors.AnonApiTimeoutError(
          "Timeout exceeded when calling GET /account/api/v1/activity.",
        );
      case "unknown":
        throw new errors.AnonApiError({
          message: _response.error.errorMessage,
        });
    }
  }

  /**
   * Create an activity event
   *
   * @param {AnonApi.CreateEventRequestBody} request
   * @param {Activity.RequestOptions} requestOptions - Request-specific configuration.
   *
   * @example
   *     await client.activity.createEvent({
   *         activityType: "link",
   *         datetime: 1,
   *         groupId: "string",
   *         data: {
   *             app: "string",
   *             type: "linkStart"
   *         },
   *         userId: "string"
   *     })
   */
  public async createEvent(
    request: AnonApi.CreateEventRequestBody,
    requestOptions?: Activity.RequestOptions,
  ): Promise<AnonApi.CreateEventResponse> {
    const _response = await core.fetcher({
      url: urlJoin(
        (await core.Supplier.get(this._options.environment)) ??
          environments.AnonApiEnvironment.Sandbox,
        "/account/api/v1/activity",
      ),
      method: "POST",
      headers: {
        Authorization: await this._getAuthorizationHeader(),
        "X-Anon-Sdk-Client-Id": await core.Supplier.get(
          this._options.anonSdkClientId,
        ),
        "X-Fern-Language": "JavaScript",
        "X-Fern-Runtime": core.RUNTIME.type,
        "X-Fern-Runtime-Version": core.RUNTIME.version,
        ...requestOptions?.headers,
      },
      contentType: "application/json",
      requestType: "json",
      body: request,
      timeoutMs:
        requestOptions?.timeoutInSeconds != null
          ? requestOptions.timeoutInSeconds * 1000
          : 60000,
      maxRetries: requestOptions?.maxRetries,
      abortSignal: requestOptions?.abortSignal,
    });
    if (_response.ok) {
      return _response.body as AnonApi.CreateEventResponse;
    }

    if (_response.error.reason === "status-code") {
      throw new errors.AnonApiError({
        statusCode: _response.error.statusCode,
        body: _response.error.body,
      });
    }

    switch (_response.error.reason) {
      case "non-json":
        throw new errors.AnonApiError({
          statusCode: _response.error.statusCode,
          body: _response.error.rawBody,
        });
      case "timeout":
        throw new errors.AnonApiTimeoutError(
          "Timeout exceeded when calling POST /account/api/v1/activity.",
        );
      case "unknown":
        throw new errors.AnonApiError({
          message: _response.error.errorMessage,
        });
    }
  }

  /**
   * Create an event to an existing activity
   *
   * @param {string} groupId - ID of the activity to add the event to
   * @param {AnonApi.CreateEventRequestBody} request
   * @param {Activity.RequestOptions} requestOptions - Request-specific configuration.
   *
   * @example
   *     await client.activity.createEventWithGroupId("string", {
   *         activityType: "link",
   *         datetime: 1,
   *         groupId: "string",
   *         data: {
   *             app: "string",
   *             type: "linkStart"
   *         },
   *         userId: "string"
   *     })
   */
  public async createEventWithGroupId(
    groupId: string,
    request: AnonApi.CreateEventRequestBody,
    requestOptions?: Activity.RequestOptions,
  ): Promise<AnonApi.CreateEventResponse> {
    const _response = await core.fetcher({
      url: urlJoin(
        (await core.Supplier.get(this._options.environment)) ??
          environments.AnonApiEnvironment.Sandbox,
        `/account/api/v1/activity/${encodeURIComponent(groupId)}`,
      ),
      method: "POST",
      headers: {
        Authorization: await this._getAuthorizationHeader(),
        "X-Anon-Sdk-Client-Id": await core.Supplier.get(
          this._options.anonSdkClientId,
        ),
        "X-Fern-Language": "JavaScript",
        "X-Fern-Runtime": core.RUNTIME.type,
        "X-Fern-Runtime-Version": core.RUNTIME.version,
        ...requestOptions?.headers,
      },
      contentType: "application/json",
      requestType: "json",
      body: request,
      timeoutMs:
        requestOptions?.timeoutInSeconds != null
          ? requestOptions.timeoutInSeconds * 1000
          : 60000,
      maxRetries: requestOptions?.maxRetries,
      abortSignal: requestOptions?.abortSignal,
    });
    if (_response.ok) {
      return _response.body as AnonApi.CreateEventResponse;
    }

    if (_response.error.reason === "status-code") {
      throw new errors.AnonApiError({
        statusCode: _response.error.statusCode,
        body: _response.error.body,
      });
    }

    switch (_response.error.reason) {
      case "non-json":
        throw new errors.AnonApiError({
          statusCode: _response.error.statusCode,
          body: _response.error.rawBody,
        });
      case "timeout":
        throw new errors.AnonApiTimeoutError(
          "Timeout exceeded when calling POST /account/api/v1/activity/{groupId}.",
        );
      case "unknown":
        throw new errors.AnonApiError({
          message: _response.error.errorMessage,
        });
    }
  }

  protected async _getAuthorizationHeader(): Promise<string> {
    return `Bearer ${await core.Supplier.get(this._options.token)}`;
  }
}
