// @anon/fern-sdk/src/api/resources/run/types/Integration.ts
/**
 * This file was auto-generated by Fern from our API Definition.
 */

export type Integration =
  /**
   * Amazon.com */
  | "amazon"
  /**
   * Delta Air Lines */
  | "delta"
  /**
   * DoorDash */
  | "doordash"
  /**
   * Facebook */
  | "facebook"
  /**
   * Github */
  | "github"
  /**
   * Instacart */
  | "instacart"
  /**
   * Instagram */
  | "instagram"
  /**
   * LinkedIn */
  | "linkedin"
  /**
   * OpenTable */
  | "opentable"
  /**
   * Twitter / X */
  | "twitter"
  /**
   * Resy */
  | "resy"
  /**
   * Uber */
  | "uber"
  /**
   * United Airlines */
  | "united_airlines"
  /**
   * Nextdoor */
  | "nextdoor"
  /**
   * Do not use */
  | "example";

export const Integration = {
  Amazon: "amazon",
  Delta: "delta",
  DoorDash: "doordash",
  Facebook: "facebook",
  Github: "github",
  Instacart: "instacart",
  Instagram: "instagram",
  LinkedIn: "linkedin",
  OpenTable: "opentable",
  Twitter: "twitter",
  Resy: "resy",
  Uber: "uber",
  United: "united_airlines",
  Nextdoor: "nextdoor",
  Example: "example",
} as const;
