// @anon/fern-sdk/dist/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnonApiClient = void 0;
const Client_1 = require("./api/resources/actions/client/Client");
const Client_2 = require("./api/resources/activity/client/Client");
const Client_3 = require("./api/resources/appIntegrations/client/Client");
const Client_4 = require("./api/resources/appUserIdToken/client/Client");
const Client_5 = require("./api/resources/link/client/Client");
const Client_6 = require("./api/resources/member/client/Client");
const Client_7 = require("./api/resources/organization/client/Client");
const Client_8 = require("./api/resources/run/client/Client");
const Client_9 = require("./api/resources/runtimeLog/client/Client");
const Client_10 = require("./api/resources/sdkClient/client/Client");
const Client_11 = require("./api/resources/session/client/Client");
const Client_12 = require("./api/resources/userPool/client/Client");
class AnonApiClient {
    _options;
    constructor(_options) {
        this._options = _options;
    }
    _actions;
    get actions() {
        return (this._actions ??= new Client_1.Actions(this._options));
    }
    _activity;
    get activity() {
        return (this._activity ??= new Client_2.Activity(this._options));
    }
    _appIntegrations;
    get appIntegrations() {
        return (this._appIntegrations ??= new Client_3.AppIntegrations(this._options));
    }
    _appUserIdToken;
    get appUserIdToken() {
        return (this._appUserIdToken ??= new Client_4.AppUserIdToken(this._options));
    }
    _link;
    get link() {
        return (this._link ??= new Client_5.Link(this._options));
    }
    _member;
    get member() {
        return (this._member ??= new Client_6.Member(this._options));
    }
    _organization;
    get organization() {
        return (this._organization ??= new Client_7.Organization(this._options));
    }
    _run;
    get run() {
        return (this._run ??= new Client_8.Run(this._options));
    }
    _runtimeLog;
    get runtimeLog() {
        return (this._runtimeLog ??= new Client_9.RuntimeLog(this._options));
    }
    _sdkClient;
    get sdkClient() {
        return (this._sdkClient ??= new Client_10.SdkClient(this._options));
    }
    _session;
    get session() {
        return (this._session ??= new Client_11.Session(this._options));
    }
    _userPool;
    get userPool() {
        return (this._userPool ??= new Client_12.UserPool(this._options));
    }
}
exports.AnonApiClient = AnonApiClient;
